File: Crio/crio_programming_interview_problems-master/CommonAncestor/CommonAncestor.java
import java.util.*;
import crio.ds.Tree.TreeNode;

// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

class Pair {
    int first;
    int second;
    Pair(int a, int b) {
        first = a;
        second = b;
    }
}

public class CommonAncestor {
    TreeNode getNode( TreeNode root , int p) {

        if ( root == null) return null;
        if ( root.val == p) return root;
        TreeNode a  = getNode(root.left, p);
        if (a != null) {
            return a;
        }
        TreeNode b = getNode(root.right, p);
        return b;
    }
    TreeNode makeTreeFromEdges(long parent[] , ArrayList<Pair> edges) {
        HashMap<Integer , TreeNode> nodebyIndex = new HashMap<Integer , TreeNode>();
        int n = parent.length;
        for (int i = 0; i < n; i++) {
            nodebyIndex.put(i, new TreeNode(parent[i]));
        }
        TreeNode root = nodebyIndex.get(0);
        for (int i = 0; i < n; i++) {
            TreeNode ptr = nodebyIndex.get(i);
            if (edges.get(i).first != -1) {

                ptr.left = nodebyIndex.get(edges.get(i).first);
            }
            if (edges.get(i).second != -1) {
                ptr.right = nodebyIndex.get(edges.get(i).second);
            }
        }
        for ( int i = 0; i < n ; i++) {
            TreeNode ptr = nodebyIndex.get(i);
            ptr.val = parent[i];
        }
    
        return root;
    }
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int tests = scanner.nextInt();
        for(int t = 0; t <  tests ; t++) {
            int n, a, b;
            n = scanner.nextInt();
            long parent[] = new long[n];
            for (int i = 0; i < n; i++) {
                parent[i] = scanner.nextLong();
            }
            ArrayList<Pair> edges = new ArrayList<Pair>();
            for (int i = 0; i < n; i++) {
                int x;
                x = scanner.nextInt(); 
                a = scanner.nextInt();
                b = scanner.nextInt();
                edges.add(new Pair(a-1, b-1));
            }
            TreeNode root = new CommonAncestor().makeTreeFromEdges(parent, edges);
            a = scanner.nextInt();
            b = scanner.nextInt();
            TreeNode p, q;
            p =  new CommonAncestor().getNode(root, a);
		    q =     new CommonAncestor().getNode(root, b);
            TreeNode answer = new Solution().commonAncestor(root , p  , q);
            System.out.println(answer.val);
        }
    }
}

File: Crio/Recursion/12_CommonAncestor/CommonAncestor.py
from Solution import commonAncestor
from crio.ds.Tree.TreeNode import *

# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

def getTreeFromInput(n, val, edges):
    nodes = []
    for i in range(n):
        nodes.append(TreeNode(val[i]))

    for edge in edges:
        parentIdx, leftChildIdx, rightChildIdx = edge

        if leftChildIdx >= 0:
            nodes[parentIdx-1].left = nodes[leftChildIdx-1]
        
        if rightChildIdx >= 0:
            nodes[parentIdx-1].right = nodes[rightChildIdx-1]

    return nodes[0]

def getNode(root, val):
    if not root:
        return None
    if root.val == val:
        return root
    left = getNode(root.left, val)
    if left:
        return left
    return getNode(root.right, val)

if __name__ == '__main__':
    import sys
    sys.setrecursionlimit(100000)

    testCases = int(input())
    for _t in range(testCases):
        n = int(input())
        val = list(map(int, input().strip().split()))
        assert(sorted(list(set(val))) == sorted(val))

        edges = []
        for i in range(n):
            edges.append(list(map(int, input().strip().split())))
        pVal, qVal = map(int, input().split())

        root = getTreeFromInput(n, val, edges)
        
        p = getNode(root, pVal)
        q = getNode(root, qVal)

        assert p and q

        result = commonAncestor(root, p, q)
        print(result.val)